### DISCUSSION TITLE
UnusedImports strict interpretation

### DISCUSSION THREAD

**Discussion by linusjf** [pmd/pmd#3130] [Q&A] [answered]
With reference to the rule [UnusedImports](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedimports), I have a static import of all methods in a class but the methods referenced in the code are from its base class. The rule flags the derived class' import as unused despite the compiler recognising the import as valid.

In my opinion, this is fine since it's (strictly) true that the methods do not belong to the specified import. 

Is this the expected behaviour?

```java

class Hello {

public static void sayHello() {

}
}

class HelloMore extends Hello {

}

import static HelloMore.*; // flagged as unused

public class Test {

public static void main(String... args) {
sayHello();
}
}
```

**Comment by oowekyala** [comment, marked as answer]
I think not. The rule should only flag imports that may be removed without changing the program (otherwise it's not unused right)

**Reply by linusjf** [reply]
I interpret it as forcing the user the use the correct import. If there are methods available in the derived class and if even one is referenced, then it is used and there is no complation or rule violation for the base class' methods. It's a corner case that probably will occur only in error.

**Reply by linusjf** [reply]
I think the rule must recognise that the parent class' methods are being referenced instead and suggest that the parent class be used instead. It's more of a case of MisusedImport rather than UnusedImport.

**Reply by linusjf** [reply]
https://github.com/pmd/pmd/issues/3132 Reported.
